zkevm-circuits/src/evm_circuit/execution/callop.rs

```diff
@@ -256,107 +301,152 @@ impl<F: Field> ExecutionGadget<F> for CallOpGadget<F> {
-                ),
-            ] {
-                cb.call_context_lookup(true.expr(), None, field_tag, value);
-            }
-
-            // Setup next call's context.
-            for (field_tag, value) in [
-                (CallContextFieldTag::CallerId, cb.curr.state.call_id.expr()),
-                (CallContextFieldTag::TxId, tx_id.expr()),
-                (CallContextFieldTag::Depth, depth.expr() + 1.expr()),
-                (CallContextFieldTag::CallerAddress, caller_address),
-                (CallContextFieldTag::CalleeAddress, callee_address),
-                (CallContextFieldTag::CallDataOffset, cd_address.offset()),
-                (CallContextFieldTag::CallDataLength, cd_address.length()),
-                (CallContextFieldTag::ReturnDataOffset, rd_address.offset()),
-                (CallContextFieldTag::ReturnDataLength, rd_address.length()),
-                (
-                    CallContextFieldTag::Value,
-                    select::expr(is_delegatecall.expr(), current_value.expr(), value.expr()),
-                ),
-                (CallContextFieldTag::IsSuccess, is_success.expr()),
-                (CallContextFieldTag::IsStatic, is_staticcall.expr()),
-                (CallContextFieldTag::LastCalleeId, 0.expr()),
-                (CallContextFieldTag::LastCalleeReturnDataOffset, 0.expr()),
-                (CallContextFieldTag::LastCalleeReturnDataLength, 0.expr()),
-                (CallContextFieldTag::IsRoot, 0.expr()),
-                (CallContextFieldTag::IsCreate, 0.expr()),
-                (CallContextFieldTag::CodeHash, callee_code_hash.expr()),
-            ] {
-                cb.call_context_lookup(true.expr(), Some(callee_call_id.expr()), field_tag, value);
-            }
-
-            // Give gas stipend if value is not zero
-            let callee_gas_left = callee_gas_left + has_value * GAS_STIPEND_CALL_WITH_VALUE.expr();
-
-            // Both CALL and CALLCODE have an extra stack pop `value`, and
-            // opcode DELEGATECALL has two extra call context lookups - current
-            // caller address and current value.
-            let rw_counter_delta =
-                43.expr() + is_call.expr() + is_callcode.expr() + is_delegatecall.expr() * 2.expr();
-            cb.require_step_state_transition(StepStateTransition {
-                rw_counter: Delta(rw_counter_delta),
-                call_id: To(callee_call_id.expr()),
-                is_root: To(false.expr()),
-                is_create: To(false.expr()),
-                code_hash: To(callee_code_hash.expr()),
-                gas_left: To(callee_gas_left),
-                reversible_write_counter: To(2.expr()),
-                ..StepStateTransition::new_context()
-            });
-        });
+        // handle is_insufficient_balance step transition
+        cb.condition(
+            is_insufficient_balance.expr() * (is_call.expr() + is_callcode.expr()),
```

Comment:
Suggestion: you could constraint `value` to be `0` when `not(is_call.expr() + is_callcode.expr())` after line 132.
This way `is_insufficient_balance` is a valid value for all opcodes, and it's enough in this condition.

Otherwise, I believe `is_insufficient_balance` depends on `value` which is free when `not(is_call.expr() + is_callcode.expr())`, so the prover can chose the value for `is_insufficient_balance` freely for some opcodes, which seems dangerous (specially since this value is used in later conditions, so the prover could skip all call context transitions in DELEGATECALL, STATICCALL by making `is_insufficient_balance=1` by assigning a very high value to `value`).

Reply:
for static call, there is already constraint in line L199: 
```
cb.condition(has_value.clone(), |cb| {
            cb.require_zero(
                "CALL with value must not be in static call stack",
                is_static.expr(),
            );
        });
```
you mean replace this constraint by  :
```
cb.condition(not(is_call.expr() + is_callcode.expr()), |cb| {
            cb.require_zero(
                "CALL with value must not be in static call stack",
                value.expr(),
            );
        });
```
and  just  `cb.condition( is_insufficient_balance.expr() ....` check insufficient balance dedicated constraints ?. @ed255

Reply:
> for static call, there is already constraint in line L199:
> 
> ```
> cb.condition(has_value.clone(), |cb| {
>             cb.require_zero(
>                 "CALL with value must not be in static call stack",
>                 is_static.expr(),
>             );
>         });
> ```

I see that this `is_static` means that the current context is static, but not that the opcode is `STATICCALL`.  I think this constraint should be kept.

> you mean replace this constraint by :
> 
> ```
> cb.condition(not(is_call.expr() + is_callcode.expr()), |cb| {
>             cb.require_zero(
>                 "CALL with value must not be in static call stack",
>                 value.expr(),
>             );
>         });
> ```
> 
> and just `cb.condition( is_insufficient_balance.expr() ....` check insufficient balance dedicated constraints ?. @ed255

I think the original constraint must be kept because it's constraining value to be 0 when the context is static.
The constraint I'm suggesting is to avoid the `value` witness to be underconstrained for `DELEGATECALL` and `STATICCALL`.  So my proposal is to **add** the constraint you mentioned, and just `cb.condition( is_insufficient_balance.expr() ....`  like you mention.

Reply:
> I think the original constraint must be kept because it's constraining value to be 0 when the context is static.
> The constraint I'm suggesting is to avoid the `value` witness to be underconstrained for `DELEGATECALL` and `STATICCALL`. So my proposal is to **add** the constraint you mentioned, and just `cb.condition( is_insufficient_balance.expr() ....` like you mention.

I see, I will try it !!

Reply:
updated in [cb71717](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/998/commits/cb717177ba6df5ef2a30824274daf5238de5cf3c), thanks for suggestion !